scripts/join/join_test.go

```diff
@@ -0,0 +1,191 @@
+	if !*integration {
+		t.Skip("skipping integration test")
+	}
+
+	const timeout = time.Hour * 6
+
+	network := netconf.Omega
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	home := t.TempDir()
+	logsPath, err := filepath.Abs(*logsFile)
+	require.NoError(t, err)
+	output, err := os.OpenFile(logsPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	require.NoError(t, err)
+
+	cfg := clicmd.InitConfig{
+		Network: network,
+		Home:    home,
+		Moniker: t.Name(),
+		HaloTag: getGitCommit7(t),
+	}
+
+	require.NoError(t, ensureHaloImage(cfg.HaloTag))
+
+	log.Info(ctx, "Exec: omni operator init-nodes")
+	require.NoError(t, clicmd.InitNodes(log.WithNoopLogger(ctx), cfg))
+
+	t0 := time.Now()
+
+	log.Info(ctx, "Exec: docker compose up", "logs_file", logsPath)
+	cmd := exec.CommandContext(ctx, "docker", "compose", "up")
+	cmd.Stderr = output
+	cmd.Stdout = output
+	cmd.Dir = home
+
+	var eg errgroup.Group
+	eg.Go(func() error {
+		// Start the nodes.
+		err := cmd.Run()
+		if err == nil || ctx.Err() != nil {
+			return nil // Docker compose didn't error
+		}
+
+		defer cancel()
+
+		return errors.Wrap(err, "docker compose up early exit")
+	})
+	eg.Go(func() error {
+		// Monitor the progress until synced.
+		ethCl, err := ethclient.Dial("omni_evm", "http://localhost:8545")
+		require.NoError(t, err)
+
+		ticker := time.NewTicker(time.Second * 30)
+		defer ticker.Stop()
+
+		timeoutCtx, timeoutCancel := context.WithTimeout(ctx, timeout)
+		defer timeoutCancel()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-timeoutCtx.Done():
+				cancel()
+				return errors.New("timed out waiting for sync", "duration", "duration", time.Since(t0).Truncate(time.Second))
+			case <-ticker.C:
+				haloStatus, err := retry(ctx, haloStatus)
+				require.NoError(t, err)
+
+				execStatus, err := retry(ctx, ethCl.SyncProgress)
+				require.NoError(t, err)
+				execSynced := execStatus.Done()
+				execHeight := execStatus.HighestBlock
+
+				log.Info(ctx, "Status",
+					"halo_status", haloStatus,
+					"execution_synced", execSynced,
+					"execution_height", execHeight,
+					"duration", time.Since(t0).Truncate(time.Second),
+				)
+
+				if haloStatus == "healthy" {
+					if !execSynced {
+						return errors.New("halo healthy but execution chain not synced", "height", execHeight)
+					}
+
+					log.Info(ctx, "Synced 🎉", "duration", time.Since(t0).Truncate(time.Second))
+					cancel()
+
+					return nil
+				}
+			}
+		}
+	})
+
+	if err := eg.Wait(); err != nil {
+		printLogsTail(t, logsPath)
+		tutil.RequireNoError(t, err)
+	}
+}
+
+func printLogsTail(t *testing.T, path string) {
+	t.Helper()
+	bz, err := os.ReadFile(path)
+	require.NoError(t, err)
+
+	lines := strings.Split(string(bz), "\n")
+	const n = 50
+	if len(lines) > n {
+		lines = lines[len(lines)-n:]
+	}
+
+	fmt.Println(strings.Join(lines, "\n"))
+}
+
+func retry[R any](ctx context.Context, fn func(context.Context) (R, error)) (R, error) {
```

Comment:
from recent usage of retry it feels this could be added to lib

Reply:
nah, not worth it